libbf.h

```diff
@@ -36,8 +36,8 @@
 #define LIMB_BITS (1 << LIMB_LOG2_BITS)
 
 #if LIMB_BITS == 64
-typedef __int128 int128_t;
-typedef unsigned __int128 uint128_t;
+__extension__ typedef __int128 int128_t;
+__extension__ typedef unsigned __int128 uint128_t;
 typedef int64_t slimb_t;
 typedef uint64_t limb_t;
 typedef uint128_t dlimb_t;
```

Comment:
Change OK as long as `__extension__` is properly defined elsewhere if required.

We should probably not define `int128_t`  and  `uint128_t` if they do not exist and use `bf_` prefixed names for all public types.  Big Floats are expected to be replaced with a simpler and faster BigNum package anyway so fixing this is not necessary.

Reply:
> Change OK as long as __extension__ is properly defined elsewhere if required.

Great! Just to clarify though, `__extension__` is a standard [compiler keyword](https://gcc.gnu.org/onlinedocs/gcc/Alternate-Keywords.html), is there a place in the doc or similar where I should define it?

> We should probably not define int128_t and uint128_t if they do not exist and use bf_ prefixed names for all public types.

According to GCC, `__int128` should exist if `long long` is [at least 128 bits wide](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fint128.html). So these definitions and their usage in `libbf.c` could be guarded by a `sizeof(long long) >= 16` conditional or similar. Is that something I should do here or save for a separate PR?

Reply:
> > We should probably not define int128_t and uint128_t if they do not exist and use bf_ prefixed names for all public types.
> 
> According to GCC, `__int128` should exist if `long long` is [at least 128 bits wide](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fint128.html). So these definitions and their usage in `libbf.c` could be guarded by a `sizeof(long long) >= 16` conditional or similar. Is that something I should do here or save for a separate PR?

`int128_t` is potentially defined in `<stdint.h>`,  I think you should use `#ifdef INT128_MAX` as a guard.  `sizeof` cannot be used in preprocessor expressions.

Reply:
> > Change OK as long as **__extension__** is properly defined elsewhere if required.
> 
> Great! Just to clarify though, `__extension__` is a standard [compiler keyword](https://gcc.gnu.org/onlinedocs/gcc/Alternate-Keywords.html), is there a place in the doc or similar where I should define it?
> 
`__extension__` is not a standard keyword, it is a reserved name.  I was unsure when it was introduced in **gcc** and **clang**. Testing with [godbolt.org](https://godbolt.org/z/5jG58r13G), I checked that it is supported since at least gcc 3.4.3 and clang 3.0, so no problem with that.

Reply:
> __extension__ is not a standard keyword, it is a reserved name. This name is used by gcc, clang and other compilers, but not msvc.

Good to know, I had no idea!

> int128_t is potentially defined in <stdint.h>, I think you should use #ifndef INT128_MAX as a guard to avoid redefining int128_t if it exists already

Do any implementations provide `int128_t`? I can only find references to `__int128` and `__int128_t`

Testing with the latest clang & gcc, the only int128-related define is `SIZEOF_INT128`:

```
% echo "#include <stdint.h>" | clang-18 -dM -E - | grep -i int128
#define __SIZEOF_INT128__ 16

% echo "#include <stdint.h>" | gcc-14 -dM -E - | grep -i int128
#define __SIZEOF_INT128__ 16
```

Apparently this is defined from [`gcc` >= 4.6 and `clang` >= 3.3](https://stackoverflow.com/a/54815033).

Reply:
Quoting the C23 standard:

> **7.22.1.1 Exact-width integer types**<br>
> 1 The typedef name `intN_t` designates a signed integer type with width `N` and no padding bits. Thus, `int8_t` denotes such a signed integer type with a width of exactly 8 bits.
> 2 The typedef name `uintN_t` designates an unsigned integer type with width `N` and no padding bits. Thus, `uint24_t` denotes such an unsigned integer type with a width of exactly 24 bits.
> 3 If an implementation provides standard or extended integer types with a particular width and no padding bits, it shall define the corresponding typedef names.

> **J.6.1 Rule based identifiers**
> 1 The following 53 regular expressions characterize identifiers that are systematically reserved by some clause in this document.
> [...] `int[a-zA-Z0-9_]*_t`, [...] `uint[a-zA-Z0-9_]*_t`

This makes `int128_t` and `uint128_t` reserved names. 

For some reason, neither **gcc** not **clang** define `int128_t` or `uint128_t` eventhough they do support 128-bit integer arithmetics, but don't accept 128-bit integer literals and the standard library does not support 128-bit `printf` or `scanf` conversions, albeit there is a way to specify these in the format string: `%w128d`, `%w128u`, etc.

Using different identifiers would avoid potential potability issues and testing for `INT128_MAX` would be a portable way to support 128-bit arithmetics on targets that have standard support for it.

Reply:
> This makes int128_t and uint128_t reserved names.

Ah apologies, I misunderstood.

For the current changes, should I also add the change to the `bf_` identifiers like below:

```c
#ifndef INT128_MAX
__extension__ typedef __int128 int128_t;
__extension__ typedef unsigned __int128 uint128_t;
#endif
typedef int128_t bf_int128_t;
typedef uint128_t bf_uint128_t;
```

Or was there a different naming scheme for the `bf_` prefix that you'd prefer? I can also handle that in a separate PR if that's easier

Reply:
Actually, only add the `#ifndef` guard. Do not bother defining the `bf_xxx` types as this would create too many changes in both **libbf.c** and **quickjs.c**.  This module is going to change to keep minimal changes.

Reply:
No worries, current changes should be good to go then